// testcontainers/dist/log-system-diagnostics.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logSystemDiagnostics = void 0;
const logger_1 = require("./logger");
const docker_compose_1 = require("./docker-compose/docker-compose");
const get_info_1 = require("./docker/functions/get-info");
const logSystemDiagnostics = (dockerode) => __awaiter(void 0, void 0, void 0, function* () {
    logger_1.log.debug("Fetching system diagnostics");
    const info = {
        node: getNodeInfo(),
        docker: yield (0, get_info_1.getDockerInfo)(dockerode),
        dockerCompose: yield getDockerComposeInfo(),
    };
    logger_1.log.debug(`System diagnostics: ${JSON.stringify(info)}`);
});
exports.logSystemDiagnostics = logSystemDiagnostics;
const getNodeInfo = () => {
    return {
        version: process.version,
        architecture: process.arch,
        platform: process.platform,
    };
};
const getDockerComposeInfo = () => __awaiter(void 0, void 0, void 0, function* () {
    try {
        return {
            version: (yield (0, docker_compose_1.version)()).data.version,
        };
    }
    catch (err) {
        logger_1.log.info(`Unable to detect docker-compose version, is it installed? ${err}`);
        return undefined;
    }
});
