// testcontainers/dist/docker/functions/get-info.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDockerInfo = void 0;
const getDockerInfo = (dockerode) => __awaiter(void 0, void 0, void 0, function* () {
    const info = yield dockerode.info();
    return {
        serverVersion: info.ServerVersion,
        operatingSystem: info.OperatingSystem,
        operatingSystemType: info.OSType,
        architecture: info.Architecture,
        cpus: info.NCPU,
        memory: info.MemTotal,
    };
});
exports.getDockerInfo = getDockerInfo;
