// testcontainers/dist/docker/docker-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dockerClient = void 0;
const dockerode_1 = __importDefault(require("dockerode"));
const path_1 = __importDefault(require("path"));
const logger_1 = require("../logger");
const url_1 = require("url");
const fs_1 = require("fs");
const run_in_container_1 = require("./functions/run-in-container");
const log_system_diagnostics_1 = require("../log-system-diagnostics");
require("../testcontainers-properties-file");
const getDockerClient = () => __awaiter(void 0, void 0, void 0, function* () {
    const strategies = [
        new ConfigurationStrategy(),
        new UnixSocketStrategy(),
        new NpipeSocketStrategy(),
    ];
    for (const strategy of strategies) {
        if (strategy.isApplicable()) {
            logger_1.log.debug(`Found applicable Docker client strategy: ${strategy.getName()}`);
            const { uri, dockerode } = yield strategy.initialise();
            logger_1.log.debug(`Testing Docker client strategy URI: ${uri}`);
            if (yield isDockerDaemonReachable(dockerode)) {
                const host = yield resolveHost(dockerode, uri);
                logger_1.log.info(`Using Docker client strategy: ${strategy.getName()}, Docker host: ${host}`);
                (0, log_system_diagnostics_1.logSystemDiagnostics)(dockerode);
                return { host, dockerode };
            }
            else {
                logger_1.log.warn(`Docker client strategy ${strategy.getName()} is not reachable`);
            }
        }
    }
    throw new Error("No Docker client strategy found");
});
const isDockerDaemonReachable = (dockerode) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const response = yield dockerode.ping();
        return response.toString() === "OK";
    }
    catch (err) {
        logger_1.log.warn(`Docker daemon is not reachable: ${err}`);
        return false;
    }
});
class ConfigurationStrategy {
    initialise() {
        return __awaiter(this, void 0, void 0, function* () {
            const { DOCKER_HOST, DOCKER_TLS_VERIFY, DOCKER_CERT_PATH } = process.env;
            const dockerOptions = {};
            const { pathname, hostname, port } = new url_1.URL(DOCKER_HOST);
            if (hostname !== "") {
                dockerOptions.host = hostname;
                dockerOptions.port = port;
            }
            else {
                dockerOptions.socketPath = pathname;
            }
            if (DOCKER_TLS_VERIFY === "1" && DOCKER_CERT_PATH !== undefined) {
                dockerOptions.ca = yield fs_1.promises.readFile(path_1.default.resolve(DOCKER_CERT_PATH, "ca.pem"));
                dockerOptions.cert = yield fs_1.promises.readFile(path_1.default.resolve(DOCKER_CERT_PATH, "cert.pem"));
                dockerOptions.key = yield fs_1.promises.readFile(path_1.default.resolve(DOCKER_CERT_PATH, "key.pem"));
            }
            return {
                uri: DOCKER_HOST,
                dockerode: new dockerode_1.default(dockerOptions),
            };
        });
    }
    isApplicable() {
        return process.env.DOCKER_HOST !== undefined;
    }
    getName() {
        return "ConfigurationStrategy";
    }
}
class UnixSocketStrategy {
    initialise() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                uri: "unix:///var/run/docker.sock",
                dockerode: new dockerode_1.default({ socketPath: "/var/run/docker.sock" }),
            };
        });
    }
    isApplicable() {
        return process.platform === "linux" || process.platform === "darwin";
    }
    getName() {
        return "UnixSocketStrategy";
    }
}
class NpipeSocketStrategy {
    initialise() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                uri: "npipe:////./pipe/docker_engine",
                dockerode: new dockerode_1.default({ socketPath: "//./pipe/docker_engine" }),
            };
        });
    }
    isApplicable() {
        return process.platform === "win32";
    }
    getName() {
        return "NpipeSocketStrategy";
    }
}
const resolveHost = (dockerode, uri) => __awaiter(void 0, void 0, void 0, function* () {
    if (process.env.TESTCONTAINERS_HOST_OVERRIDE !== undefined) {
        return process.env.TESTCONTAINERS_HOST_OVERRIDE;
    }
    const { protocol, hostname } = new url_1.URL(uri);
    switch (protocol) {
        case "http:":
        case "https:":
        case "tcp:":
            return hostname;
        case "unix:":
        case "npipe:": {
            if (isInContainer()) {
                const gateway = yield findGateway(dockerode);
                if (gateway !== undefined) {
                    return gateway;
                }
                const defaultGateway = yield findDefaultGateway(dockerode);
                if (defaultGateway !== undefined) {
                    return defaultGateway;
                }
            }
            return "localhost";
        }
        default:
            throw new Error(`Unsupported protocol: ${protocol}`);
    }
});
const findGateway = (dockerode) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c;
    logger_1.log.debug(`Checking gateway for Docker host`);
    const inspectResult = yield dockerode.getNetwork("bridge").inspect();
    return (_c = (_b = (_a = inspectResult === null || inspectResult === void 0 ? void 0 : inspectResult.IPAM) === null || _a === void 0 ? void 0 : _a.Config) === null || _b === void 0 ? void 0 : _b.find((config) => config.Gateway !== undefined)) === null || _c === void 0 ? void 0 : _c.Gateway;
});
const findDefaultGateway = (dockerode) => __awaiter(void 0, void 0, void 0, function* () {
    logger_1.log.debug(`Checking default gateway for Docker host`);
    return (0, run_in_container_1.runInContainer)(dockerode, "alpine:3.14", ["sh", "-c", "ip route|awk '/default/ { print $3 }'"]);
});
const isInContainer = () => (0, fs_1.existsSync)("/.dockerenv");
let _dockerClient;
const dockerClient = () => {
    if (!_dockerClient) {
        _dockerClient = getDockerClient();
    }
    return _dockerClient;
};
exports.dockerClient = dockerClient;
