// testcontainers/dist/docker-compose/default-docker-compose-options.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultDockerComposeOptions = void 0;
const defaultDockerComposeOptions = (_a) => {
    var { env = {} } = _a, options = __rest(_a, ["env"]);
    return {
        log: false,
        cwd: options.filePath,
        config: options.files,
        composeOptions: options.composeOptions,
        commandOptions: options.commandOptions,
        env: Object.assign(Object.assign(Object.assign({}, process.env), { COMPOSE_PROJECT_NAME: options.projectName }), env),
    };
};
exports.defaultDockerComposeOptions = defaultDockerComposeOptions;
